aggregator/handler/common/traffic_controller.go

```diff
@@ -0,0 +1,184 @@
+
+	"go.uber.org/atomic"
+)
+
+const (
+	defaultDefaultDisabled = false
+	defaultInitTimeout     = 2 * time.Second
+)
+
+// TrafficControlOptions configurates the traffic control options.
+type TrafficControlOptions interface {
+	// SetStore sets the kv store.
+	SetStore(store kv.Store) TrafficControlOptions
+
+	// Store returns the kv store.
+	Store() kv.Store
+
+	// SetDefaultDisabled sets the default disabled value.
+	SetDefaultDisabled(value bool) TrafficControlOptions
+
+	// DefaultDisabled returns the default disabled value.
+	DefaultDisabled() bool
+
+	// SetRuntimeDisableKey sets the runtime disable key.
+	SetRuntimeDisableKey(value string) TrafficControlOptions
+
+	// RuntimeDisableKey returns the runtime disable key.
+	RuntimeDisableKey() string
+
+	// SetInitTimeout sets the init timeout.
+	SetInitTimeout(value time.Duration) TrafficControlOptions
+
+	// InitTimeout returns the init timeout.
+	InitTimeout() time.Duration
+
+	// SetInstrumentOptions sets the instrument options.
+	SetInstrumentOptions(value instrument.Options) TrafficControlOptions
+
+	// InstrumentOptions returns the instrument options.
+	InstrumentOptions() instrument.Options
+}
+
+type trafficControlOptions struct {
+	store             kv.Store
+	defaultDisabled   bool
+	runtimeDisableKey string
+	initTimeout       time.Duration
+	instrumentOpts    instrument.Options
+}
+
+// NewTrafficControlOptions creats a new TrafficControlOptions.
+func NewTrafficControlOptions() TrafficControlOptions {
+	return &trafficControlOptions{
+		defaultDisabled: defaultDefaultDisabled,
+		initTimeout:     defaultInitTimeout,
+		instrumentOpts:  instrument.NewOptions(),
+	}
+}
+
+func (opts *trafficControlOptions) SetStore(store kv.Store) TrafficControlOptions {
+	o := *opts
+	o.store = store
+	return &o
+}
+
+func (opts *trafficControlOptions) Store() kv.Store {
+	return opts.store
+}
+
+func (opts *trafficControlOptions) SetDefaultDisabled(value bool) TrafficControlOptions {
+	o := *opts
+	o.defaultDisabled = value
+	return &o
+}
+
+func (opts *trafficControlOptions) DefaultDisabled() bool {
+	return opts.defaultDisabled
+}
+
+func (opts *trafficControlOptions) SetRuntimeDisableKey(value string) TrafficControlOptions {
+	o := *opts
+	o.runtimeDisableKey = value
+	return &o
+}
+
+func (opts *trafficControlOptions) RuntimeDisableKey() string {
+	return opts.runtimeDisableKey
+}
+
+func (opts *trafficControlOptions) SetInitTimeout(value time.Duration) TrafficControlOptions {
+	o := *opts
+	o.initTimeout = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InitTimeout() time.Duration {
+	return opts.initTimeout
+}
+
+func (opts *trafficControlOptions) SetInstrumentOptions(value instrument.Options) TrafficControlOptions {
+	o := *opts
+	o.instrumentOpts = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InstrumentOptions() instrument.Options {
+	return opts.instrumentOpts
+}
+
+// TrafficController controls if traffic is enabled.
+type TrafficController struct {
```

Comment:
Have you considered using the runtime manager in `runtime` package for controlling the runtime behavior? This is an established pattern in this repo where the objects wishing to listen to runtime config changes register with the runtime manager and take actions on updates, and the runtime manager manages watching the keys and updating the watchable to send notifications to the watchers, which is a very clean divide of responsibility.

Reply:
Agreed it's more involved than watching a single key but I think it's not too difficult if you give a unique ID to each backend (e.g., "ingester-short-static" for the legacy one and "ingester-short-dynamic" for the new one). Then you can set one key per backend and distinguish the keys among the backends similarly to how the runtime manager keeps track of keys.

I guess it's fine to stick w/ the kv based approach if this is going to live on a branch.

Reply:
Sure, I thought about that, but that makes the runtime value update logic rather complex/unituitive, we'd have to define runtime configs for some of the backends at the aggregator level, far away from the backend configuration itself which is not great. 
I guess the main difference here is that the current keys in the runtime options are expected for the aggregator, but the traffic control logic is optional per backend.

Reply:
Yeah I've been meaning to move the handler outside the aggregator package since the interfaces are cleanly defined and separated at the moment. After that, you could leave the existing runtime manager for the aggregator as is (maybe move it into the aggregator package), and create a new runtime manager just for the handlers (maybe keep it in the handler package), in which case it'll make more sense to configure the backend keys for the handler runtime manager.